### PR TITLE
add github style search for Packages

### PR DESCRIPTION
**Pull request by madhur-tandon**
Linting would probably fail because the regex is long. And changing it so that it is multiple lines kind of changes the regex if I recall correctly. So this can be ignored for now.

### CODE REVIEW

File: quetz/utils.py [hunk 1]
```diff
@@ -24,6 +30,74 @@ def add_entry_for_index(files, subdir, fname, data_bytes):
     )
 
 
+def parse_query(query):
+    accepted_filters = ['channel', 'description', 'summary']
+    query = unquote(query.strip())
```
Comment (btel) [inline quetz/utils.py:35]: do you need it? I would assume that fastapi would already do it for you

File: quetz/utils.py [hunk 2]
```diff
@@ -24,6 +30,74 @@ def add_entry_for_index(files, subdir, fname, data_bytes):
...
+    regex = r"""(\S+:'(?:[^'\\]|\\.)*')|(\S+:\"(?:[^\"\\]|\\.)*\")|(-?\"(?:[^\"\\]|\\.)*\")|(-?'(?:[^'\\]|\\.)*')|\S+|\S+:\S+"""
```
Comment (btel) [inline quetz/utils.py:36]: impressive! perhaps you might use named groups to make it more legible

also make sure that, you can skip any of the query terms. For example, you might search only by channel name:

```
channel=conda-forge,my-channel
```

it's possible that you already do this, i don't know

File: quetz/main.py [hunk 3]
```diff
@@ -744,15 +744,16 @@ def delete_package_version(
...
 def search(
-    query: str,
+    q: str,
     dao: Dao = Depends(get_dao),
     auth: authorization.Rules = Depends(get_rules),
 ):
     user_id = auth.get_user()
-    return dao.search_packages(query, user_id)
+    keywords, filters = parse_query(q)
+    return dao.search_packages(keywords, filters, user_id)
```
Comment (btel) [inline quetz/main.py:756]: we should also make sure that the user has permission to see the packages from the search results (based on channel and package permission)

File: quetz/utils.py [hunk 4]
```diff
@@ -24,6 +30,74 @@ def add_entry_for_index(files, subdir, fname, data_bytes):
     )
 
 
+def parse_query(query):
+    accepted_filters = ['channel', 'description', 'summary']
```
Comment (btel) [inline quetz/utils.py:34]: how about filtering by user (owner)?

**Review by btel — COMMENTED** [review]
it looks very good. make sure that you cover all possible variations in the tests (see above for one example). Another thing to check is whether you can add arbitary number of whitespace between the query terms. This should give the same result:

`channel=conda-forge description=' test package'`

`channel=conda-forge      description=' test package'`

also whitespaces at the beginning and end should be allowed. 

I think that using a parser might simplify a bit the regex stuff. Have a look at the parser that I sent you (https://ply.readthedocs.io/en/latest/ply.html#lex-example) or different one, and check whether it might help. At the end, you are constructing here a DSL (domain-specific language,  Also have a look at this blog post: https://medium.com/swlh/how-to-write-your-own-domain-specific-language-in-python-a36b72d3e36c

File: quetz/utils.py (hunk 4 above)
Comment (wolfv) [reply quetz/utils.py:34]: or uploader. although that is more tricky as that is part of the packageversion data.

File: quetz/utils.py (hunk 2 above)
Comment (wolfv) [reply quetz/utils.py:36]: yeah a bit more documentation on this regex would be great :)

also maybe we could write a small parser for this mini-language, that's not based on regex?

File: quetz/utils.py (hunk 1 above)
Comment (madhur-tandon) [reply quetz/utils.py:35]: Yes, we are mimicking of what we would ideally receive from text entered into a search box. We can anyway remove this later if we figure out that FastAPI does this OR perhaps if the content entered into a search box stays the same.

File: quetz/utils.py (hunk 2 above)
Comment (madhur-tandon) [reply quetz/utils.py:36]: Yes, it is not necessary to supply all the arguments. It's up to us how many we supply. I did give a demo of this today.
I partially borrowed this regex (and did some of my own modifications) so I will have to see how it works at a deeper level as well.

I can check docs about named groups. And, I need to investigate about making a parser a bit more.

File: quetz/utils.py (hunk 4 above)
Comment (madhur-tandon) [reply quetz/utils.py:34]: Yes, we are gonna expand functionality one by one. Right now, all things listed in `PackageSearch` class in `rest_models.py` are considered.

File: quetz/utils.py (hunk 2 above)
Comment (wolfv) [reply quetz/utils.py:36]: @madhur-tandon maybe using the Python tokenizer is even easier? It splits out "strings" with ticks already, and it's part of the standard library.

https://docs.python.org/3/library/tokenize.html

I am sure there are other tokenizer libraries implemented in Python around. That might be easier & more maintainable than this regex, in the long term :)

Comment (madhur-tandon) [reply quetz/utils.py]: We are now using `shlex` (https://docs.python.org/3/library/shlex.html) as suggested by @wolfv

File: quetz/utils.py (hunk 4 above)
Comment (madhur-tandon) [reply quetz/utils.py]: Added `uploader` as well

**Pull request merged**